magento/magento2
# Fix/styles priority media mobile mixin for extend file

### Description (*)
In the first implementation of the styles architecture in Magento 2, there were imported 3 files general files:

- `_module.less`
- `_widgets.less` (doesn't matter in this case)
- `_extend.less`

and only after them, the `_responsive.less` file has been imported:
- https://github.com/magento/magento2/blob/1cd6c174ab13ee0616ccff9bae502c16f670e14c/app/design/frontend/Magento/blank/web/css/styles-m.less
- https://github.com/magento/magento2/blob/1cd6c174ab13ee0616ccff9bae502c16f670e14c/app/design/frontend/Magento/blank/web/css/styles-l.less

but it was changed in the commit https://github.com/magento/magento2/commit/605b0212bd49f04895a769de128c05b6d10c54cf:
- https://github.com/magento/magento2/blob/605b0212bd49f04895a769de128c05b6d10c54cf/app/design/frontend/Magento/blank/web/css/styles-m.less#L45
- https://github.com/magento/magento2/blob/605b0212bd49f04895a769de128c05b6d10c54cf/app/design/frontend/Magento/blank/web/css/styles-l.less#L45

The `_extend.less` file has been moved to the end of both files. It looks like a minor change and probably it wasn't checked well, but for many years the media mobile styles inside the `max` mixin are broken for all the `_extend.less` files.

Let me show you 2 examples.

#### _module.less

```less
& when (@media-common = true) {
    body {
        &::before {
            content: '_module.less common';
        }
    }
}

.media-width(@extremum, @break) when (@extremum = 'max') and (@break = @screen__m) {
    body {
        &::before {
            content: '_module.less mobile mixin';
        }
    }
}

```

In this style structure we're expecting the next priority (from less to higher):
1. common styles
2. media query mixin

And we really have the expected result:
<img width="1541" alt="image" src="https://github.com/magento/magento2/assets/28815318/3ae5b7c6-79e2-40da-953b-99df3584e456">

#### _extend.less

But let's try to do that with the `_extend.less` file. Just rename the file from `_module.less` to `_extend.less`. Add minor changes to the content:

```less
& when (@media-common = true) {
    body {
        &::before {
            content: '_extend.less common';
        }
    }
}

.media-width(@extremum, @break) when (@extremum = 'max') and (@break = @screen__m) {
    body {
        &::before {
            content: '_extend.less mobile mixin';
        }
    }
}
```

And re-compile styles again. Check the result:
<img width="1369" alt="image" src="https://github.com/magento/magento2/assets/28815318/bc7db320-482e-4905-b24a-d2aa3d78e9fd">

As you can see, the priority has changed just by the renaming. 

The reason is the wrong order of the imported files in `styles-m.less` and `styles-l.less` files. The `_responsive.less` files and variables have to be always imported at the end of the files, especially after the `_extend.less` files.

So, my PR has to repair the expected logic, that media queries have to always have the higher priority, 'cause they're written after the common styles.

### Manual testing scenarios (*)
1. Choose the main branch.
2. Create a new theme.
3. Apply the theme in the admin panel.
4. Create a new styles file in the theme:
```
app/design/frontend/Vendor/theme/web/css/source/_module.less
```
5. Add the content from the description above for the `_module.less` file.
6. To make the result more clear, try to use the `empty` layout for the home page. Or just create a new file: 
```
app/design/frontend/Vendor/theme/Magento_Theme/layout/default.xml
```
With the content:
```xml
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd"
>
    <update handle="empty"/>
</page>

```
7. Clear all the static files and cache or just run the Grunt LESS compilation.
8. Check the result on the storefront. You have to see the same as in my first screenshot above.
9. Then rename the file from `_module.less` to `_extend.less` with the same content. Just rename the file name output inside.
10. Re-compile styles again.
11. Check the result.
12. Common styles have a higher priority then styles from the mobile query mixin.
13. Apply changes from the current PR.
15. Re-compile styles.
16. Check again, issue has to be fixed.

### Contribution checklist (*)


### Resolved issues:
1. [x] resolves magento/magento2#38136: Fix/styles priority media mobile mixin for extend file
